### PR TITLE
Return null from ClaimsPrincipal.Current if no thread principal.

### LINKED ISSUE #42581: ClaimsPrincipal.CurrentPrincipal throws ArgumentNullException if no principal is set.
This is on .NET 5 RC1 Windows

Repo:

Create a new console app and insert

``` csharp
class Program
{
	static void Main(string[] args)
	{
		System.Security.Claims.ClaimsPrincipal.Current?.ToString();
	}
}
```

Expected behavior: Runs without error

Actual behavior:

```
Unhandled exception. System.ArgumentNullException: Value cannot be null. (Parameter 'principal')
   at System.Security.Claims.ClaimsPrincipal..ctor(IPrincipal principal)
   at System.Security.Claims.ClaimsPrincipal.SelectClaimsPrincipal()
   at System.Security.Claims.ClaimsPrincipal.get_Current()
```

### PR DESCRIPTION
**Pull request by vcsjones**
In the .NET Framework, the AppDomain's default PrincipalPolicy is
UnauthenticatedPrincipal. When falling back to the thread's principal,
this would no throw because a GenericPrincipal was returned.
If the policy was changed to NoPrincipal, ClaimsPrincipal.Current would
throw constructing the ClaimsPrincipal since there is no principal.
However that behavior was opt-in.

In .NET Core, the default policy was changed to NoPrincipal, which would
mean unless a principal was explicitly set, Current would always throw.

This changes the behavior to return null instead of throw for
NoPrincipal. This is to continue supporting the existing behavior in
.NET Core 3 of returning null if there is no principal, but continue
to allow falling back to the thread's principal.

Fixes #42581

### CODE REVIEW

File: src/libraries/System.Security.Claims/src/System/Security/Claims/ClaimsPrincipal.cs [hunk 1]
```diff
@@ -27,9 +27,19 @@ private enum SerializationMask
...
+            // UnauthenticatedPrincipal. In .NET Core, the default is NoPrincipal. .NET Framework
+            // would throw an ArgumentNullException when constructing the ClaimsPrincipal with a
+            // null principal from the thread if it were set to use NoPrincpial. In .NET Core, since
+            // NoPrinicipal is the default, we return null instead of throw.
+
+            return Thread.CurrentPrincipal switch {
+                ClaimsPrincipal claimsPrincipal => claimsPrincipal,
+                not null => new ClaimsPrincipal(Thread.CurrentPrincipal),
+                null => null
+            };
```
Comment (stephentoub) [inline src/libraries/System.Security.Claims/src/System/Security/Claims/ClaimsPrincipal.cs:42]:
```suggestion
            IPrincipal? p = Thread.CurrentPrincipal;
            return p switch {
                ClaimsPrincipal claimsPrincipal => claimsPrincipal,
                not null => new ClaimsPrincipal(p),
                null => null
            };
```

**Review by stephentoub — APPROVED** [review]

**Comment by jeffhandley** [comment]
Tagging @eerhardt for review too

**Review by jeffhandley — APPROVED** [review]
Thanks, @vcsjones!

**Review by bartonjs — APPROVED** [review]

**Review by eerhardt — APPROVED** [review]

**Comment by vcsjones** [comment]
@jeffhandley did you mean to tag the other PR?

**Comment by jeffhandley** [comment]
@vcsjones I did. 😊 I've updated the other as well.  Great work on this fix; it got approved for Ask Mode in 5.0 RC2.  Thank you!

**Pull request merged**